sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/routing/LocationCache.java

```diff
@@ -272,6 +272,23 @@ public boolean shouldRefreshEndpoints(Utils.ValueHolder<Boolean> canRefreshInBac
             return false;
         }
     }
+
+    public String getRegionName(URI locationEndpoint, com.azure.cosmos.implementation.OperationType operationType) {
+        Utils.ValueHolder<String> regionName = new Utils.ValueHolder<>();
+        if (operationType.isWriteOperation()) {
+            if (Utils.tryGetValue(this.locationInfo.regionNameByWriteEndpoint, locationEndpoint, regionName)) {
```

Comment:
does the new method public String getRegionName(URI locationEndpoint, operationType) relying on the internal state of the LocationCache guarantee that always give accurate region name in any scenario?

I would like you to think about any edge cases :-)

Reply:
discussed with Naveen offline. We agreed there is an edge case race condition where this can give incorrect result when the SDK prespective on the region topology changes. One potential solution for this is to keep UriWithRegion as pair (URI, region) in the globalEndpointManager. because this was more involved work we agreed to leave this for future.